File: scilab/modules/elementary_functions/src/fortran/idmin.f
      integer function idmin(n, x, incx)
*
*     PURPOSE
*        finds the index of the first element having minimum value
*        without taking into account the nan(s)
*
*     NOTE
*        - original version modified by Bruno (for the nan problem...)
*          (01/01/2003)
*        - this function return 1 if x has only nan components : may be
*          this is not a good behavior
*        - this function doesn't test if n<1 or incx<1 : this is done
*          by the scilab interface
*
      implicit none
      integer n, incx
      double precision x(incx,*)

      double precision xmin
      integer i, j
      external isanan
      integer  isanan

      idmin = 1

*     initialize the min with the first component being not a nan
      j = 1      
      do while ( isanan(x(1,j)) .eq. 1 )
         j = j + 1
         if ( j .gt. n ) return
      enddo
      xmin = x(1,j)
      idmin = j

*     the usual loop
      do i = j+1, n
         if ( x(1,i) .lt. xmin) then  ! a test with a nan must always return false
            xmin = x(1,i)
            idmin = i
         endif
      enddo
 
      end

File: scilab/modules/elementary_functions/src/c/idmin.c
/*----------------------------------------------------------------------*/
/* @author INRIA (initial fortran version)                              */
/* @author Bruno Pincon (2003) bug fix for the nan problem              */
/* @author Allan Cornet (2007) rewrite to C                             */
/*----------------------------------------------------------------------*/
#include "idmin.h"
#include "isanan.h"
/*--------------------------------------------------------------------------*/ 
int C2F(idmin)(int *n, double *x, int *incx)
{

    int x_dim1 = 0, x_offset  = 0, ret_val = 0, i1 = 0;
    int i = 0, j = 0;
    double xmin  = 0;
    
    x_dim1 = *incx;
    x_offset = 1 + x_dim1;
    x -= x_offset;

    ret_val = 1;
	/* initialize the min with the first component being not a nan */
    j = 1;
    while(C2F(isanan)(&x[j * x_dim1 + 1]) == 1) 
	{
		++j;
		if (j > *n) return ret_val;
    }
    xmin = x[j * x_dim1 + 1];
    ret_val = j;

	/* the usual loop */
    i1 = *n;
    for (i = j + 1; i <= i1; ++i) 
	{
		if (x[i * x_dim1 + 1] < xmin) 
		{
			/* a test with a nan must always return false */
			xmin = x[i * x_dim1 + 1];
			ret_val = i;
		}
    }
    return ret_val;
} 
/*--------------------------------------------------------------------------*/ 
